### PR TITLE
Add Dockerfile and Github Action

### PR DESCRIPTION
**Pull request by katenels**
This will allow us to build docker images, as well as publish docker images to dockerhub.

### CODE REVIEW

**Review by adzuci — COMMENTED** [review]
This looks good in general, but I think we need to resolve the codecov warnings before merging, cc @jdmulloy

File: registrar/docker_gunicorn_configuration.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+
+def close_all_caches():
+    """
+    Closes the cache so that newly forked workers cannot accidentally
+    share the socket with the processes they were forked from.
+    Prevents a race condition in which one worker could get a cache response
+    intended for another worker.
+
+    Safe for 1.4 and 1.8.
```
Comment (matthugs) [inline registrar/docker_gunicorn_configuration.py]: I was curious what version numbers we were referring to here. [Based on this](https://github.com/edx/configuration/commit/fa033cc0d57d62d0471e499df3aa5ff1aabcfe24) it looks like these were django versions. We definitely don't use either of those versions of django and haven't for a long time, so I wonder whether or not we need this still?

**Comment by ohnickmoy** [comment]
To resolve those warnings down the line, you can place unittests [in the apps/core/tests part of this repo.](https://github.com/edx/registrar/tree/master/registrar/apps/core/tests)

**Review by ohnickmoy — APPROVED** [review]
To resolve those warnings down the line, you can place unittests [in the apps/core/tests part of this repo.](https://github.com/edx/registrar/tree/master/registrar/apps/core/tests). 

unit test docs for python can be found [here](https://docs.python.org/3/library/unittest.html)

File: registrar/docker_gunicorn_configuration.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+    Args:
+        worker (Worker): The worker to process the request
+        req (Request): The request to be processed
+
+    Returns:
+        None
+    """
+    worker.log.info("%s %s" % (req.method, req.path))
+
```
Comment (ohnickmoy) [inline registrar/docker_gunicorn_configuration.py:26]: @adzuci, is this just printing out log statements, does the req needs to be mocked? https://docs.python.org/3/library/unittest.mock.html

**Review by ohnickmoy — APPROVED** [review]
To resolve those warnings down the line, you can place unittests [in the apps/core/tests part of this repo.](https://github.com/edx/registrar/tree/master/registrar/apps/core/tests). 

unit test docs for python can be found [here](https://docs.python.org/3/library/unittest.html)

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+WORKDIR /edx/app/registrar
+
+# Copy just Python requirements & install them.
+COPY requirements/ /edx/app/registrar/requirements/
+COPY Makefile /edx/app/registrar/
+RUN make production-requirements
+
+# Code is owned by root so it cannot be modified by the application user.
+# So we copy it before changing users.
+USER app
```
Comment (kdmccormick) [inline Dockerfile]: Seems like we copy the code later. Can you fix the comment or fix the order of the commands?

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+# Python is Python3.
+RUN ln -s /usr/bin/pip3 /usr/bin/pip
+RUN ln -s /usr/bin/python3 /usr/bin/python
+
+# Use UTF-8.
+RUN locale-gen en_US.UTF-8
+ENV LANG en_US.UTF-8
+ENV LANGUAGE en_US:en
+ENV LC_ALL en_US.UTF-8
+ENV DJANGO_SETTINGS_MODULE registrar.settings.production
```
Comment (kdmccormick) [inline Dockerfile:27]: Style nit: this seems separate from the block of UTF-8-related commands.
```suggestion

ENV DJANGO_SETTINGS_MODULE registrar.settings.production
```

**Review by kdmccormick — COMMENTED** [review]
Cool; nice work!

A couple comments below. Please also add a `.dockerignore` file to make sure things like `.git/`, `registrar/settings/private.py`, etc. are excluded from the image. [This article](https://codefresh.io/docker-tutorial/not-ignore-dockerignore-2/) does a good job explaining the concept and rationale.

**Review by matthugs — APPROVED** [review]

File: registrar/docker_gunicorn_configuration.py (hunk 1 above)
Comment (katenels) [reply registrar/docker_gunicorn_configuration.py]: I removed that line from the docstring but I don't know enough about this to remove the second part of this method, so I've left it in for now

File: registrar/docker_gunicorn_configuration.py (hunk 2 above)
Comment (adzuci) [reply registrar/docker_gunicorn_configuration.py:26]: It's just printing the logs, but to test it we would need to create a sample worker and request object with a method and path which could be passed into `pre_request`.  I don't think we would need to mock the actual request for this method though.

**Review by adzuci — APPROVED** [review]
:shipit:

**Pull request merged**